### PR TITLE
[None][models] Avoid GEMM in Gemma4 vision RoPE

### PR DESCRIPTION
**Pull request by 2ez4bz**
## Dev Engineer Review

The source change replaces GEMM with elementwise rotary-frequency multiplication. This protects phase precision when TF32 is enabled. The change may alter performance, so benchmark the vision RoPE path if performance is a concern. No API changes are reported.

## QA Engineer Review

`tests/unittest/_torch/modeling/test_modeling_gemma4.py` adds a regression test for Gemma4 vision RoPE generation. The test rejects `aten.mm` and `aten.bmm` dispatch and compares cosine and sine outputs with a double-precision reference on a 64×64 grid. No integration test-list changes are shown. Coverage verdict: **sufficient**. Test execution results are unavailable.

## Per-File QA Perspective

- `tensorrt_llm/_torch/models/modeling_gemma4_vision.py`: Verify RoPE numerical accuracy and confirm that rotary-table generation does not dispatch GEMM operators. Check for unacceptable performance regression.
- `tests/unittest/_torch/modeling/test_modeling_gemma4.py`: Covers the TF32 precision regression and forbidden matrix-multiply dispatch. The unit test does not require an integration test-list entry.

## Description

* Why?

TF32-enabled matrix multiplication can reduce rotary phase precision for the vision position embeddings.

* What?

Compute the rotary outer product elementwise and add a regression test that verifies no matrix-multiply operators are dispatched.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tests/unittest/_torch/modeling/test_modeling_gemma4.py [hunk 1]
```diff
@@ -4268,6 +4269,40 @@ def test_vision_attention_uses_native_rmsnorm_adapter(self):
...
+
+        with _RejectMatrixMultiply():
+            cos, sin = rotary(torch.empty(0), position_ids)
+
+        inv_freq = rotary.inv_freq.double()[None, None, :]
+        reference_parts = [
+            inv_freq * position_ids[:, :, dim].double().unsqueeze(-1) for dim in range(2)
+        ]
+        reference = torch.cat([torch.cat((part, part), dim=-1) for part in reference_parts], dim=-1)
+        torch.testing.assert_close(cos.double(), reference.cos(), atol=2e-6, rtol=0)
```
Comment (brnguyen2) [inline tests/unittest/_torch/modeling/test_modeling_gemma4.py]: `atol=2e-6` is at the fp32 error floor for this grid, not a real margin. The largest phases land in the `[32, 64)` binade (`inv_freq[1] ≈ 0.5995`, `p ∈ 54..63` → ~37.8), where half-ulp of float32 is `2^-19 ≈ 1.91e-6`; add `~1` ulp from `cosf` (1.19e-7 for `|cos| ∈ [0.5,1)`) and the worst case is ≈2.03e-6. It passes only because those ~10 products happen not to round near half-ulp — change the grid to 128 and `inv_freq[1]·127 ≈ 76` moves into `[64,128)` where half-ulp is 3.8e-6, and the test fails for reasons unrelated to GEMM.

Suggest `atol=1e-5`. TF32 truncates operand mantissas to 10 bits (`2^-11` relative), which is ~0.03 rad of phase error at `p=63`, so `1e-5` still rejects the regression by three orders of magnitude.

Separately: building the reference from `rotary.inv_freq.double()` reuses the module's own fp32 table, which is the right call here (an independently-computed fp64 `inv_freq` would blow any sane tolerance), but it does mean `inv_freq` construction itself isn't validated by this test.

**Review by brnguyen2 — APPROVED** [review]
Approving — the comments below are optional touch-ups, not blockers.

Two things beyond the inline note.

`tensorrt_llm/_torch/auto_deploy/models/custom/modeling_gemma4.py:739` has the same rank-1 `@` outer product in the AutoDeploy copy of `Gemma4VisionRotaryEmbedding`, so the AutoDeploy path keeps whatever exposure this PR removes from the PyTorch path. Not this PR's scope, but worth a follow-up so the two implementations don't drift.

Was the precision loss observed, or is this preventive? Nothing in the repo sets `torch.backends.cuda.matmul.allow_tf32` or `set_float32_matmul_precision` outside a scoped enable in the DSA indexer, and PyTorch defaults matmul TF32 off — so under TRT-LLM's own defaults this bmm was already full fp32 and the change only matters if an embedding application flips the global knob. If you did measure a real accuracy drop somewhere, linking it would help a future reader; if it's hardening, saying so in the description is enough.

**Review by aswinvisva — APPROVED** [review]
